### DISCUSSION TITLE
How to make the q-page-container bounded by the ui.footer?

### DISCUSSION THREAD

**Discussion by JS-Aibel** [zauberzeug/nicegui#1314] [Q&A] [answered]
### Question

In the example below the table is overlapping the ui.footer. This is due to the h-screen class on the ui.aggrid.
How can I set the table or the q-page-container itself to be bounded by the top of the footer?

```py
from nicegui import ui
import pandas as pd

@ui.page('/')
def page_layout():
    with ui.header(elevated=True):
        ui.label('HEADER')
    with ui.footer():
        ui.label('FOOTER')
    
    ui.label('CONTENT')
    df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
    ui.aggrid.from_pandas(df).classes('h-screen')

ui.run()
```

**Comment by falkoschindler** [comment]
Interesting problem, @JS-Aibel!

Here is a possible solution:
```py
with ui.header():
    ui.label('HEADER')
with ui.footer():
    ui.label('FOOTER')
with ui.column().classes('w-full h-[calc(100vh-53px-32px-53px)]'):
    ui.label('CONTENT')
    ui.aggrid({}).classes('flex-grow')
```

Let me explain:
- We wrap the main content in a `ui.column` spanning the whole width and a height calculated from the following components:
  - 100vh is 100% of the view height
  - 53px is the header height
  - 32px accounts for the padding around the main content
  - 53px is the footer height
- We let the AG Grid grow to automatically fill the remaining space.

![localhost_1234__ (1)](https://github.com/zauberzeug/nicegui/assets/5767091/3e572cf4-b0f4-4394-ad21-e7d8242627bd)

**Reply by JS-Aibel** [reply]
Thank you for the answer, @falkoschindler. 
This will work OK.

**I have some follow up questions:**
Is there a way to find the pixel height of the components? For cases where the header and footer are not 53px.

In the example below one can see that the q-page-container has automatically set the padding style to 53px for top and bottom. Why will the use of ‘flex-grow’, not automatically bound the table height by the header and footer? Since the paddig size is defined in its’ grandparent (q-page-container) component.


![image](https://github.com/zauberzeug/nicegui/assets/100705556/f1ebc42e-1519-4cfa-9ff2-9694307cc2aa)

```py
from nicegui import ui

@ui.page('/')
def page_layout():
    with ui.header():
        ui.label('HEADER')
    with ui.footer():
        ui.label('FOOTER')

    ui.aggrid({}).classes('flex-grow')

ui.run()
```

**Reply by falkoschindler** [reply]
Yes, hard-coding the header and footer height is not nice. But determining the height via JavaScript calls wouldn't be a great solution either. And there's probably no way to determine it without asking the client.
But I found another solution for your problem. See below. 🙂

**Comment by falkoschindler** [comment, marked as answer]
I just found a solution manipulating the overall page layout. By setting the `.nicegui-content` element to full width and the `.q-page` element to flex layout, the content automatically grows to fill the vertical space. Using `flex-grow` for the AG Grid table, it grows to fill the space of the content:

```py
ui.query('.nicegui-content').classes('w-full')
ui.query('.q-page').classes('flex')

with ui.header():
    ui.label('HEADER')
with ui.footer():
    ui.label('FOOTER')

ui.label('CONTENT')
ui.aggrid({}).classes('flex-grow')
```

**Reply by JS-Aibel** [reply]
Perfect, I will try to use this.
Thank you.
